// @plunk/node/dist/lib/Plunk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Plunk = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const request_1 = require("../utils/request");
class Plunk {
    constructor(key) {
        /**
         * Resources
         */
        this.events = {
            /**
             * Publishes an event to Plunk
             * @param {string} event.event - The event you want to publish
             * @param {string} event.email - The email associated with this event
             * @param {Object=} event.data - The user data associated with this event
             */
            track: (event) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                return yield (0, request_1.request)(this.client, {
                    method: "POST",
                    url: "/track",
                    body: Object.assign({}, event),
                });
            }),
        };
        this.emails = {
            /**
             * Sends a transactional email with Plunk
             *
             * @param {string} body.to - The email you want to send to
             * @param {string} body.subject - The subject of the email
             * @param {string} body.body - The body of the email
             * @param {string=} body.from - The email you want to send from
             * @param {string=} body.name - The name you want to send as
             * @param {boolean=false} body.withUnsubscribe - Whether to include an unsubscribe link
             */
            send: (body) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                return yield (0, request_1.request)(this.client, {
                    method: "POST",
                    url: "/send",
                    body: Object.assign({}, body),
                });
            }),
        };
        this.key = key;
        this.client = axios_1.default.create({
            baseURL: "https://api.useplunk.com/v1",
            headers: {
                Authorization: `Bearer ${this.key}`,
            },
        });
    }
}
exports.Plunk = Plunk;
